File: NeetCode/Arrays_and_Hashing/Sort_An_Array/sort__an__array.py
from typing import Optional, List
import heapq
from collections import defaultdict
import random


class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        def mergeS(array):
            if len(array) <= 1:
                return array
            mid = len(array) // 2
            left = mergeS(array[:mid])
            right = mergeS(array[mid:])
            return merge(left, right)

        def merge(left, right):
            result = []
            i, j = 0, 0
            while i < len(left) and j < len(right):
                if left[i] <= right[j]:
                    result.append(left[i])
                    i += 1
                else:
                    result.append(right)
                    j += 1

            result.extend(left[i:])
            result.extend(right[j:])
            return result

        return mergeS(nums)


if __name__ == "__main__":
    sol = Solution()
    nums = [10, 9, 1, 1, 1, 2, 3, 1]
    print(sol.sortArray(nums))
    print("Running Solution...")


""" 
works but fails on leetcode 
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        def swap(i, j):
            nums[i], nums[j] = nums[j], nums[i]

        def partition(low, high):
            pivot_index = random.randint(low, high)
            swap(pivot_index, high)
            pivot = nums[high]
            i = low - 1
            for j in range(low, high):
                if nums[j] < pivot:
                    i += 1
                    swap(i, j)
            swap(i + 1, high)
            return i + 1

        def quick(low, high):
            if low < high:
                pivot = partition(low, high)
                quick(low, pivot - 1)
                quick(pivot + 1, high)

        quick(0, len(nums) - 1)
        return nums
"""

File: NeetCode/Arrays_and_Hashing/Sort_An_Array/Sort_An_Array.cpp
// Sort_An_Array.cpp
//
// Compile locally with:
// clang++ Sort_An_Array.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> sortArray(vector<int> &nums)
    {
        return mergeSort(nums);
    }
    vector<int> mergeSort(vector<int> &arr)
    {
        if (arr.size() <= 1)
        {
            return arr;
        }
        int mid = arr.size() / 2;
        vector<int> left(arr.begin(), arr.begin() + mid);
        vector<int> right(arr.begin() + mid, arr.end());
        left = mergeSort(left);
        right = mergeSort(right);
        return merge(left, right);
    }
    vector<int> merge(vector<int> &left, vector<int> &right)
    {
        vector<int> result;
        int i(0), j(0);
        while (i < left.size() && j < right.size())
        {
            if (left[i] <= right[j])
            {
                result.push_back(left[i]);
                i++;
            }
            else
            {
                result.push_back(right[j]);
                j++;
            }
        }

        while (i < left.size())
        {
            result.push_back(left[i]);
            i++;
        }

        while (j < right.size())
        {
            result.push_back(right[j]);
            j++;
        }

        return result;
    }
};

int main()
{
    cout << "Running Sort_An_Array.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
